### ISSUE TITLE
A referral-only root zone (almost)

### ISSUE THREAD

**Issue by chjj** [handshake-org/hsd#125]
After a long conversation with @mikedamm, he made a good case for making the root zone referral-only, at least in terms of DNS. It's an idea I had considered... I think right after I decided was fed up with the [complexity] of the resource serialization.

Putting non-referrals directly at the root is _odd_, and will likely confuse a lot of clients.

Note that this doesn't prevent anyone from putting arbitrary data at the root, or putting an A record directly at the apex of their zone: I could still put an A record directly at `chjj.` for example (by running my own nameserver). It may violate some RFCs, but I could do it.

I like this idea for its simplicity, and it mirrors the existing root zone.

It also allows us to rip out a ton of the resource serialization code, much of which is _not_ well thought out.

### Proposal for DNS Handshake Resources (V1)

- `DS` - A committed DS record.
- `NS` - A plain NS record with no glue (cannot point to a name under your own zone).
- `GLUE` - A name/ip pair which generates an NS record in the authority section and A/AAA records in the additional section (_must_ reside under your own zone -- this avoids DNS cache poisoning attacks, which are probably currently possible on testnet).
- `SYNTH` - An ipv4/ipv6 address which acts as a nameserver. This generates a _synthetic_ name for the NS record on the DNS side (similar to what we currently do if no name is provided). The IP itself appears in the additional section as glue.
- Other records up in the air.

This makes the JSON and binary serialization super simple. Example:

``` js
[
  {
    "type": "NS",
    "name": "ns1.chjjdns.example.com." // <-- must point to a another zone
  },
  {
    "type": "GLUE",
    "name": "ns1.chjj.", // <-- must point to our own zone
    "addr": [
      "127.0.0.1",
      "::1"
    ]
  },
  {
    "type": "SYNTH",
    "addr": "127.0.0.1" // <-- generates `_[base32]._hns.` with glue of 127.0.0.1
  },
  {
    "type": "DS",
    "keyTag": 1234,
    "algorithm": 8,
    "digestType": 2,
    "digest": "0000000000000000000000000000000000000000000000000000000000000000"
  }
]
```

Generates:

``` zone
;; AUTHORITY SECTION:
chjj. 86400 IN NS ns1.chjjdns.example.com.
chjj. 86400 IN NS ns1.chjj.
chjj. 86400 IN NS _ab7qaaab._hns.
chjj. 86400 IN DS 1234 8 2 0000000000000000000000000000000000000000000000000000000000000000
chjj. 86400 IN RRSIG DS 8 2 86400 ...

;; ADDITIONAL SECTION:
ns1.chjj. 86400 IN A 127.0.0.1
ns1.chjj. 86400 IN AAAA ::1
_ab7qaaab._hns. 86400 IN A 127.0.0.1
```

Furthermore, because of how we organized the data, we probably don't even need label compression for on-chain storage anymore.

Clients should return NXDOMAIN on any name that doesn't follow the NS/GLUE rules. Not doing so opens yourself up to DNS cache poisoning attacks. In fact, I'm considering not even serializing the final label in `glue.name`. The same zone should be _implied_.

The final incompatibility between us and (for lack of a better word) _stupid_ DNS clients is that `_ab7qaaab._hns` is not actually resolvable on ICANN's system. For whatever reason, certain clients may fallback to their default root hints when resolving HNS nameservers. We can account for this by adding a `--synthetic-nameserver` switch to HSD and HNSD. Optionally, a user can pass `--synthetic-nameserver=hns-ip-translator.example.com.`, changing `_ab7qaaab._hns` to `_ab7qaaab.hns-ip-translator.example.com.`. A public translator on ICANN's system can account for these final compatability errors. This name server would decode the lowest label as base32, and serve either an A or AAAA record in response. This is centralized, but when resolving HNS names with proper software, nobody should ever actually hit this centralized service. It's merely there so buggy DNS clients work with HNS.

### Proposal for Data Handshake Resources (V1)

Arbitrary data is _unspecified_ and we leave it to the community to come up with various serialization formats.

Clients should return NXDOMAIN on any name with arbitrary data (i.e. unparse-able).

### Final Thoughts

So, I imagine there will be some criticism of this idea. A lot of people are currently putting A records directly on chain, but there's no actual benefit to having uncensorable A records -- at the end of the day, you have to be running a server somewhere. If you can run a webserver, you can surely run a nameserver with a DNSSEC setup. HNS probably shouldn't act as your own personal nameserver.

TXT, TLSA, SMIMEA, SSHFP, etc. records are a different story. Though it may violate some RFCs, there _is_ a benefit to having uncensorable _data_ on-chain, keys in particular. I think we should discuss these record types and consider including them directly on-chain.

I will say, I _do_ like the idea of temporarily including A/AAA records during the bootstrapping process and deprecating them later. Just because it's fun and easy to setup (as opposed to setting up your own nameserver).

Opening this up for discussion.

[complexity]: https://github.com/handshake-org/hnsd/blob/master/src/resource.c

**Comment by smcki012** [comment]
This definitely makes sense given backwards compatibility and reduction in complexity -- especially if the inclusion of A/AAA records can be included initially just so things are bit easier for new users and developers as they start to tinker, then moving those outside to owned zones in future.

**Comment by pinheadmz** [comment]
@chjj Would this be enforced in any way? From what I can tell, resource serialization is not consensus and any arbitrary data (up to 512 bytes) can be included in a REGISTER covenant.

**Comment by turbomaze** [comment]
A couple thoughts:

1) I agree with Steven that allowing A records at the root makes it easier for people to play around with Handshake. Longer term, setting up a nameserver isn't that big of a deal, but it's an additional point of friction that could discourage developers from experimenting in the early days.

2) I expect that people will want to get a very good Handshake name for themselves and use it for mutliple things. Allowing both TXT records and NS records at the root simplifies future schemes that want to store data on Handshake names. People will want to store a BTC address at `btc.turbomaze/` while simultaneously serving a personal site at `turbomaze/` for example. The integrity/authenticity guarantees this gives you would still be achievable in the world where you need to go through a nameserver, but doing so is much higher friction.

**Comment by chjj** [comment]
@pinheadmz, correct. This is not consensus. We could call it policy or "layer 2 consensus" maybe. It's easier to change than a lot of on-chain stuff.

@turbomaze, agreed. I think NS, TXT, and to-be-deprecated A/AAAA records should be allowed. I don't think any serious domain would actually A/AAAA records like that in production, but it is fun for tinkering. In the future for strict DNS conformance, we could look at moving data records to some kind of pseudo-tld which is pre-reserved for that kind of thing. Say if I publish TXT records, they appear under `chjj._meta.` or something. We have a dynamic root so we can pretty much do anything. I think @mikedamm also had some thoughts on this.

If we disallow AAAA/MX records at the root, it just occurred to me that there's no reason to allow TLSA/SSHFP/SMIMEA/etc at the root either, since they work in tandem with one another.

So new v1 outline:

- DS
- NS (ns/glue/synthetic)
- TXT (possible future migration to a "meta" tld)
- A/AAAA (for future deprecation)

I think I like it. Dead simple.

I also think all records should have a default TTL controlled by the implementation. Our tree only updates every few hours, so there's not really a reason to allow custom TTLs. This could also be dynamically calculated by block height.

**Comment by stp-ip** [comment]
In my view adding A/AAAA into the mix especially with a deprecation plan is problematic. That's bloating the chain with data, that we won't want in the future anyway.

Just for easier on-boarding I'd say that's a bad compromise. We can do a centralized workaround such as a small nameserver that accepts signed messages to add/change/delete records for a given zone.
That's useful in the long term and keeps chain state growth low.

In my view DS and NS records should be enough. Let's rebuild the root zone first and then iterate on fixing the rest of the issues such as CAs, "centralized" DNS record hosting etc.

Now to the rational:

The main value add I see in handshake is creating a trustless, open and secure root zone. So optimally instead of hardcoding specific root servers into software the software would have a handshake resolved build in and can look up the root records themselves. For this to be feasible the chain data needs to stay small and easy to verify the current UTXO set aka root records.

Even censorship resistant A/AAAA records is hard to justify as one can provide their own nameserver, which is using IP, which in turn is being used by the service trying to be censorship resistance. So BGP routing is still not censorship resistant enough to make a valid case adding this in the root DNS.

What's the specific use case for allowing "unbounded" TXT records? In my view that will lead to weird spamming issues and chain bloat.
If the idea is for censorship resistant metadata, that might be a reason, but DNS might not be suited by default to provide this. Working around by using TXT records seems weird.
We could use the TXT record for distribution of censorship resistant endpoint data such as "here is my onion address" or such. Problem is this is still unbounded and should either be limited to x records per zone or be specifically expensive.
Maybe having a ._meta zone as a sidechain or the like might be a good iteration later on.

My first thoughts on the matter. Happy to discuss this in more detail.

**Comment by tynes** [comment]
Thinking about it more and more, I'm thinking that restricting the contents of the chain's records to NS and DNSSEC related records makes the most sense. The trust will then rely on the security of zones which requires good opsec.  Tools that make it easy to set up and maintain DNSSEC in a zone in an easy to audit way could help out here. I think that this approach would lead to more innovation happening at the level of DNS, which is beneficial for the internet as a whole.

Building something like Lightning's [DNS Bootstrapping](https://github.com/lightningnetwork/lightning-rfc/blob/master/10-dns-bootstrap.md) wouldn't be able to scale fully on chain since each name can only store 512 bytes.

> We can do a centralized workaround such as a small nameserver that accepts signed messages to add/change/delete records for a given zone.

I agree with this idea and think that a small "personal server" that runs a nameserver would be a good step towards making Handshake easy to build on.

**Comment by pinheadmz** [comment]
Related to this: I have a handshake name (`zippy`) with a resource that contains both a CNAME and a TXT:
```
{
  "version": 0,
  "compat": false,
  "ttl": 64,
  "canonical": "matthewzipkin.com.",
  "text": [
    "OpenBazaar: QmeSyTRaNZMD8ajcfbhC8eYibWgnSZtSGUp3Vn59bCnPWC"
  ]
}
```
When I dig @ my local hsd full node I can retrieve both data depending on the request:
```
$ dig @127.0.0.1 -p 15350  zippy TXT 
...
zippy.                  64      IN      TXT     "OpenBazaar: QmeSyTRaNZMD8ajcfbhC8eYibWgnSZtSGUp3Vn59bCnPWC"  
...
```

However, the same request made to `hnsd` does the "correct" thing and follows the CNAME automatically, ignoring the on-hsd-chain TXT entry, and instead returning the TXT entry from the referred server `matthewzipkin.com`:

```
;; ANSWER SECTION:
zippy.                  20      IN      CNAME   matthewzipkin.com.
matthewzipkin.com.      3563    IN      TXT     "v=spf1 a mx ptr include:secureserver.net ~all"
matthewzipkin.com.      3563    IN      TXT     "google-site-verification=0hsIOizsCi6NTAR8o_ew_v7i4SGMRmSaBfmDqBWlREE"
```

h/t @boymanjor for pointing out this restriction:
https://en.wikipedia.org/wiki/CNAME_record#Restrictions

**Comment by Thann** [comment]
> at the end of the day, you have to be running a server somewhere. 

I fundamentally disagree with this premise. 

1. If I want to point my record at a "github pages" site.
2. Future Torrent or IPFS specific record types would not need centralized servers to host content.
3. Identity verification.

There are many usecases for DNS that don't involve "running a server" and many more that would benefit greatly from a super highly available DNS records served directly from the chain.  For these reasons I think it's important to keep the DNS layer extensable AND able to work standalone =]

**Comment by jacobhaven** [comment]
> 1. If I want to point my record at a "github pages" site.
> 2. Future Torrent or IPFS specific record types would not need centralized servers to host content.
> 3. Identity verification.
> 
> There are many usecases for DNS that don't involve "running a server" and many more that would benefit greatly from a super highly available DNS records served directly from the chain. For these reasons I think it's important to keep the DNS layer extensable AND able to work standalone =]

Points 2 and 3 would be a TXT record of some sort, here, right? Point 1 being a CNAME.

**Comment by boymanjor** [comment]
@chjj what is the current thinking on this thread? It seems like we have soft consensus on including DS, NS (ns/glue/synth), & TXT records. I don't like the idea of including A/AAAA records only to deprecate later. The deprecation process would not be guaranteed to even occur, let alone occur smoothly. I would prefer we either keep them or get rid of them.

>We can do a centralized workaround such as a small nameserver that accepts signed messages to add/change/delete records for a given zone.

I tend to agree with the above, but I am not opposed to including A/AAAA records. I will appeal to more experienced authorities.

>I also think all records should have a default TTL controlled by the implementation. Our tree only updates every few hours, so there's not really a reason to allow custom TTLs. This could also be dynamically calculated by block height.

I think that dynamically calculated TTLs make sense given the static tree root interval.

--

I'd like to get the ball rolling on finalizing the specification so that I can help out with the implementation.

**Comment by tynes** [comment]
After digging deeper into the `Resource` serialization, I've come to the conclusion that unless the `Resource` version becomes part of consensus, then it is impossible to restrict what sorts of records are kept on chain. If v0 is deprecated for v1, then nothing is stopping somebody from using a fork of the codebase to add back v0. This would just mean that users running the software that only knows v1 will not be able to resolve names that are encoded as v0. This is wasteful for the users, since they have downloaded and validated that data.

The `Resource` version is a U8, so there are up to 256 different values that can be represented (versions are relatively scarce). I am thinking about each version as its own "application", since it can define its own serialization/deserialization methods on available data. Note that two different ways of serializing DNS records counts as two different applications under this model. Creating a new application is 100% permissionless, since there are no consensus rules around the version. Technically, applications could have nothing to do with DNS and could be accessed using an alternative protocol, like RPC. This makes me worried that there could be conflicts over a version number if 2 different groups roll out their own encodings in parallel without communicating. To make these different applications useful, the resolvers need to be running the software that understands how to deserialize the particular version correctly. If a particular fork has a version that is not supported by the "main" client, then a network effect of users that demand that version would be a convincing reason to add support to the "main" client. I am not sure of the correct definition of "main" in this context, but am going with the software that the most users run.

There are some things that I like about the fact that the `Resource` version number is not part of consensus. The free market can decide which serializations are most useful and can evolve over time to serve a changing world while the base protocol can stay the same. This creates economic opportunity for different businesses to build their applications on Handshake, bringing more money, developers and potentially users into the ecosystem. Since it is unknown how often people will update their `Resources`, buy names or send value (all of these require sending a transaction), the expected demand for blockspace is unknown. Allowing for different applications will increase the transaction throughput, which helps to secure the network by increasing its security budget through increasing the profit margin of miners with the tx fees. Having lots of transaction throughput is particularly useful when the block subsidy disappears/becomes negligible. This is a *huge* problem for Bitcoin, and is outside of the Overton Window to discuss for most Bitcoiners, as 21 million is a sacred number that must never change. In reality, it is unknown if transaction fees alone could secure the network. See this paper [A model for Bitcoin’s security and the declining block subsidy](https://uncommoncore.co/wp-content/uploads/2019/10/A-model-for-Bitcoins-security-and-the-declining-block-subsidy-v1.06.pdf) by [hasu](https://twitter.com/hasufl), [Prestwich](https://twitter.com/_prestwich) and [Brandon Curtis](https://twitter.com/bcmakes) for an analysis of possible futures when there is no block subsidy.

I am open to the possibility of adding in the `Resource` version to consensus, but then it becomes a governance problem. Who is in charge of updating the `Resource` versions? Can they be soft-forked into the protocol? Do miners need to bip9 signal for them to be activated, or do the developers control them being added? What is the complexity of writing and reviewing the code?

All of this means that Handshake can be used as a general data availability layer for any sort of application. Depending on your philosophy, this could be a good or bad thing. **I think that the focus should be on being a decentralized DNS root zone**, but allowing for different applications could help to build the network effect. Imagine a new Counterparty like system, built on Handshake. An idea that I have been exploring is solving the PoS weak subjectivity problem using Handshake PoW. You can only be guaranteed to sync a PoS chain correctly if you can download a checkpoint from an external source that is trusted and then sync through that checkpoint. The checkpoint continuously moves forward as the PoS chain extends. Where will all the PoS nodes sync these checkpoints from? It is possible to add a `Resource` type that allows for a PoS checkpoint to be stored at a specific name. This would make PoS blockchains depend on the Handshake network for security (lol), assuming that there is not another sufficient way to sync these checkpoints. In reality, it is probably cheaper to just trust some centralized organization's reputation instead of having to spend HNS to update `Resources`, but all it takes is one major accident and then it may become worth it for the PoS chains to adopt using PoW for their checkpointing mechanism.

---

tldr; I do not think that this issue is a blocker for mainnet unless `Resource` versions are added to consensus. In it's current form, anybody will be able to add their own `Resource` types in a permissionless way, so if v0 (which includes many different types of DNS records) is removed in favor of v1 (which includes far fewer types of DNS records), then somebody could easily add back in v0 in a competing client implementation and put whatever records they want on chain.

**Comment by turbomaze** [comment]
Even if we commit to a particular set of record types, people could use steganographic techniques to encode non-DNS payloads in the Handshake namestate. It would be slightly more expensive per-byte than DNS payloads since there would be an efficiency hit, but that doesn't seem too significant.

For this reason, I like Mark's proposal of supporting multiple serialization formats, indexed by the version number. I'm imagining some SLIP44-esque registry where people can gather to agree on which serializations are supported. 

One worry I have with this proposal though is if there ends up being multiple incompatible serializations for DNS. E.g. if V0 is referral-only and V1 is not, then when we talk to the purists that feel strongly about a referral-only root, Handshake no longer has a consistent story.

**Comment by tynes** [comment]
Any new serialization code will break the tests written [here](https://github.com/handshake-org/hsd/blob/master/test/dns-test.js). Happy to help update these tests to apply to any new serialization code that is pushed. The `resource-v0` vectors live [here](https://github.com/handshake-org/hsd/blob/master/test/data/resources-v0.json).

**Comment by pinheadmz** [comment]
Closing due to mainnet launch :-)
